### PR TITLE
Reorganize version header generation code

### PR DESCRIPTION
**Pull request by jskeet**
This is now desirable for both REST and gRPC APIs, so it's in the
Google.Api.Gax package, and is renamed to VersionHeaderBuilder as
it's not building a User-Agent header.

(Most of the changes are just renames.)

### CODE REVIEW

File: src/Google.Api.Gax.Grpc/ApiServerStreamingCall.cs [hunk 1]
```diff
@@ -66,7 +66,7 @@ public AsyncServerStreamingCall<TResponse> Call(TRequest request, CallSettings p
         internal ApiServerStreamingCall<TRequest, TResponse> WithUserAgent(string userAgent) =>
```
Comment (chrisdunelm) [inline src/Google.Api.Gax.Grpc/ApiServerStreamingCall.cs]: userAgent -> versionHeader

File: src/Google.Api.Gax.Grpc/MetadataExtensions.cs [hunk 2]
```diff
@@ -26,7 +26,7 @@ public static void Add(this Metadata metadata, IEnumerable<Metadata.Entry> items
         public static Metadata WithUserAgent(this Metadata metadata, string userAgent)
```
Comment (chrisdunelm) [inline src/Google.Api.Gax.Grpc/MetadataExtensions.cs]: WithVersionHeader

File: src/Google.Api.Gax.Rest/UserAgentHelper.cs [hunk 3]
```diff
@@ -6,8 +6,7 @@
  */
```
Comment (chrisdunelm) [inline src/Google.Api.Gax.Rest/UserAgentHelper.cs]: Should this class name now be changed to VersionHeaderHelper?

File: src/Google.Api.Gax/AssemblyInfo.cs [hunk 4]
```diff
@@ -10,3 +10,5 @@
...
 [assembly:InternalsVisibleTo("Google.Api.Gax.Tests,PublicKey=0024000004800000940000000602000000240000525341310004000001000100bfa2d20e82583e5f3a47efa518d8a7cda58d6e1b61f51ed3de3900f238fb01e0ca95b3ed1c68147e4865a336b252f070d5207e653c9014d271fc829c18981d1c80d418d300545c3ad0e1ff4d77c596177dbfe0a588fff57306f1d6ba655de3cfa520ad777ea73bb7a92fd6027071c94bb12d5871fcd076299de0502bca12e3aa")]
 [assembly:InternalsVisibleTo("Google.Api.Gax.Grpc,PublicKey=0024000004800000940000000602000000240000525341310004000001000100bfa2d20e82583e5f3a47efa518d8a7cda58d6e1b61f51ed3de3900f238fb01e0ca95b3ed1c68147e4865a336b252f070d5207e653c9014d271fc829c18981d1c80d418d300545c3ad0e1ff4d77c596177dbfe0a588fff57306f1d6ba655de3cfa520ad777ea73bb7a92fd6027071c94bb12d5871fcd076299de0502bca12e3aa")]
+[assembly: InternalsVisibleTo("Google.Api.Gax.Rest,PublicKey=0024000004800000940000000602000000240000525341310004000001000100bfa2d20e82583e5f3a47efa518d8a7cda58d6e1b61f51ed3de3900f238fb01e0ca95b3ed1c68147e4865a336b252f070d5207e653c9014d271fc829c18981d1c80d418d300545c3ad0e1ff4d77c596177dbfe0a588fff57306f1d6ba655de3cfa520ad777ea73bb7a92fd6027071c94bb12d5871fcd076299de0502bca12e3aa")]
```
Comment (chrisdunelm) [inline src/Google.Api.Gax/AssemblyInfo.cs]: Slightly confused as to why internals are visible to a non-test assembly?

File: test/Google.Api.Gax.Rest.Tests/UserAgentHelperTest.cs [hunk 5]
```diff
@@ -5,6 +5,8 @@
  * https://developers.google.com/open-source/licenses/bsd
  */
```
Comment (chrisdunelm) [inline test/Google.Api.Gax.Rest.Tests/UserAgentHelperTest.cs]: (And this test might want to change name to VersionHeaderHelperTest)

File: src/Google.Api.Gax.Rest/UserAgentHelper.cs (hunk 3 above)
Comment (jskeet) [inline src/Google.Api.Gax.Rest/UserAgentHelper.cs]: Not sure. It still has code to use with the User-Agent as well... unlike the gRPC one.

File: src/Google.Api.Gax.Grpc/MetadataExtensions.cs (hunk 2 above)
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/MetadataExtensions.cs]: Gah, thanks.

File: src/Google.Api.Gax/AssemblyInfo.cs (hunk 4 above)
Comment (jskeet) [inline src/Google.Api.Gax/AssemblyInfo.cs]: We want to be able to get at this from Google.Api.Gax.Rest, but not from other code. Note how the previous line is similar for Grpc, although I'm not sure which internal type that's related to.

File: src/Google.Api.Gax.Grpc/ApiServerStreamingCall.cs (hunk 1 above)
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/ApiServerStreamingCall.cs]: Done. There was another one too...

File: src/Google.Api.Gax.Rest/UserAgentHelper.cs (hunk 3 above)
Comment (jskeet) [inline src/Google.Api.Gax.Rest/UserAgentHelper.cs]: Have updated docs to include "and related headers" which I think is okay :) (And it means we don't need to make breaking changes outside GAX)

**Comment by jskeet** [comment]
Right - I've now done a search for "gent" to make sure it doesn't occur anywhere we don't want it :)

**Pull request merged**